core/src/test/java/kafka/test/ClusterInstance.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package kafka.test;
+
+import kafka.network.SocketServer;
+import kafka.test.annotation.ClusterTest;
+import org.apache.kafka.clients.admin.Admin;
+import org.apache.kafka.common.network.ListenerName;
+
+import java.util.Collection;
+import java.util.Optional;
+import java.util.Properties;
+
+public interface ClusterInstance {
+
+    enum ClusterType {
+        Zk,
+        // Raft
+    }
+
+    /**
+     * Cluster type. For now, only ZK is supported.
+     */
+    ClusterType clusterType();
+
+    /**
+     * The cluster configuration used to create this cluster. Changing data in this instance through this accessor will
+     * have no affect on the cluster since it is already provisioned.
+     */
+    ClusterConfig config();
+
+    /**
+     * The listener for this cluster as configured by {@link ClusterTest} or by {@link ClusterConfig}. If
+     * unspecified by those sources, this will return the listener for the default security protocol PLAINTEXT
+     */
```

Comment:
A cluster can have multiple listeners, right?  Is this intended to be the listener that should be used by clients?  If so, calling it the client listener makes sense.

(Technically we could have multiple client listeners too, but that's much less common)

Many clusters run with 3 listeners:
* client
* inter-broker NON-REPLICATION
* replication

and we are going to add a 4th:
* controller

Reply:
Yes, multiple listeners can be defined. This is something to solve/fix in this new framework. Only one security protocol and listener are exposed here in order to stay similar to how KafkaServerTestHarness works. It defines one security protocol and listener as members:

```
protected def securityProtocol: SecurityProtocol = SecurityProtocol.PLAINTEXT
protected def listenerName: ListenerName = ListenerName.forSecurityProtocol(securityProtocol)
``` 

Of course, more listeners can be defined through configuration, but this one is exposed by the test harness and used by many tests.

Reply:
How about naming these "clientSecurityProtocol", "clientListenerName", etc. to reflect the fact that the clients will be using these security protocols, listener names, etc.  Then we can add more listeners later for specific tests without doing a big rename.